tensorflow/core/kernels/mkl/onednn_nn_ops_benchmark.cc

```diff
@@ -0,0 +1,158 @@
+      ->set_opt_level(OptimizerOptions::L0);
+  test::Benchmark("cpu", g, &opts, nullptr, nullptr, "",
+                  /*old_benchmark_api*/ false)
+      .Run(state);
+  state.SetItemsProcessed(shape.num_elements() * state.iterations());
+  state.SetLabel(label);
+}
+
+/*
+ * The following trick for number of variadic macro arguments is taken from:
+ * http://groups.google.com/group/comp.std.c/browse_thread/thread/77ee8c8f92e4a3fb/346fc464319b1ee5?pli=1
+ * The macro PP_NARG returns the number of arguments that have been passed to
+ * it.  Currrently, it is prepared to support upto 5 arguments. It is straight
+ * forward to extend to a higher number of arguments.
+ */
+#define PP_NARG(...) PP_NARG_(__VA_ARGS__, PP_RSEQ_N())
+#define PP_NARG_(...) PP_ARG_N(__VA_ARGS__)
+#define PP_ARG_N(_1, _2, _3, _4, _5, N, ...) N
+#define PP_RSEQ_N() 5, 4, 3, 2, 1, 0
+
+// For a tensor shape {a, b, c, d} we want to produce a token a_b_c_d
+#define CONCAT_DIMS1(a) _##a
+#define CONCAT_DIMS2(a, b) _##a##_##b
+#define CONCAT_DIMS3(a, b, c) _##a##_##b##_##c
+#define CONCAT_DIMS4(a, b, c, d) _##a##_##b##_##c##_##d
+#define CONCAT_DIMS5(a, b, c, d, e) _##a##_##b##_##c##_##d##_e
+#define JOIN(x, y) JOIN_AGAIN(x, y)
+#define JOIN_AGAIN(x, y) x##y
+
+// Wrapping BENCHMARK to get benchmark name appropriate, so that the argument
+// expansion takes place before the expansion of BENCHMARK.
+#define WRAP_BENCHMARK(FUNC) BENCHMARK(FUNC)
+
+#define BM_oneDNN_Softmax(dtype, num_threads, label, ...)                    \
+  static void JOIN(BM_oneDNN_Softmax_##dtype##_intraop_##num_threads##_dims, \
+                   JOIN(CONCAT_DIMS, PP_NARG(__VA_ARGS__))(__VA_ARGS__))(    \
+      ::testing::benchmark::State & state) {                                 \
+    BM_oneDNN_Softmax<dtype>(state, {__VA_ARGS__}, num_threads, label);      \
+  }                                                                          \
+  WRAP_BENCHMARK(                                                            \
+      JOIN(BM_oneDNN_Softmax_##dtype##_intraop_##num_threads##_dims,         \
+           JOIN(CONCAT_DIMS, PP_NARG(__VA_ARGS__))(__VA_ARGS__)))            \
+      ->UseRealTime()
+
+#define BM_Eigen_Softmax(dtype, num_threads, label, ...)                       \
+  static void JOIN(BM_Eigen_Softmax_##dtype##_intraop_##num_threads##_dims,    \
+                   JOIN(CONCAT_DIMS, PP_NARG(__VA_ARGS__))(__VA_ARGS__))(      \
+      ::testing::benchmark::State & state) {                                   \
+    BM_Eigen_Softmax<dtype>(state, {__VA_ARGS__}, num_threads, label);         \
+  }                                                                            \
+  WRAP_BENCHMARK(JOIN(BM_Eigen_Softmax_##dtype##_intraop_##num_threads##_dims, \
+                      JOIN(CONCAT_DIMS, PP_NARG(__VA_ARGS__))(__VA_ARGS__)))   \
+      ->UseRealTime()
+
+// Benchmark with oneDNN library.
+BM_oneDNN_Softmax(float, 4, "float32_BERT_batch_size_1", 1, 16, 384, 384);
```

Comment:
Maybe add another macro that defines both the oneDNN and Eigen benchmark at once? E.g.,
```
#define BM_Softmax(dtype, num_threads, label, ...) \
    BM_oneDNN_Softmax(...); \
    BM_Eigen_Softmax(...); 
```

This way, we don't have to update two places when we want to edit / add shapes.

Reply:
Done. Thanks for the nice suggestions!